components/brave_new_tab_ui/hooks/PrefHookManager.ts

```diff
@@ -0,0 +1,104 @@
+import { useCallback, useEffect, useState } from 'react'
+
+export type PrefListener<OnType, KeyType extends keyof OnType> = (name: KeyType, newValue: OnType[KeyType], oldValue?: OnType[KeyType]) => void
+
+/**
+ * A class to help wrapping mojom pref access in a React hook.
+ */
+export class PrefHookManager<T> {
+    #listeners: Map<keyof T, Array<PrefListener<T, keyof T>>> = new Map()
+    #lastValue?: T
+    #savePref: (key: keyof T, value: T[keyof T]) => void
```

Comment:
TIL `T[keyof T]` will match the correct value type based on the specific `keyof T` passed in the `key` argument? Great if so.

Reply:
Nah so in this case the type definition isn't perfect. I think the type definition for that would be 

```ts
<K extends keyof T>(key: K, value: T[K]) => void
```

As is, the type definition is 'any key of T with any value of T' which I originally thought would be fine because it's only really used inside the class. However, as the fix is pretty straightforward (and because I love perfect typings) I've updated it.